File: 2020/01_ReportRepair/report.py
# ======================================================================
# Report Repair
#   Advent of Code 2020 Day 01 -- Eric Wastl -- https://adventofcode.com
#
# Python implementation by Dr. Dean Earl Wright III
# ======================================================================

# ======================================================================
#                         r e p o r t . p y
# ======================================================================
"A solver for the Advent of Code 2020 Day 01 puzzle"

# ----------------------------------------------------------------------
#                                                                 import
# ----------------------------------------------------------------------

# ----------------------------------------------------------------------
#                                                              constants
# ----------------------------------------------------------------------

# ======================================================================
#                                                                 Report
# ======================================================================


class Report(object):   # pylint: disable=R0902, R0205
    "Object for Report Repair"

    def __init__(self, text=None, part2=False):

        # 1. Set the initial values
        self.part2 = part2
        self.text = text
        self.numbers = set()

        # 2. Process text (if any)
        if text is not None and len(text) > 0:
            for number in text:
                self.numbers.add(int(number))

    def get_pair(self, total):
        "Get a pair of numbers that add to the total"
        for num in self.numbers:
            if (total - num) in self.numbers:
                return num, total - num
        return None, None

    def get_trio(self, total):
        "Get a trio of numbers that add to the total"
        for num in self.numbers:
            pair = self.get_pair(total - num)
            if pair[0] is not None:
                return [num, pair[0], pair[1]]
        return None, None, None

    def part_one(self, verbose=False, limit=0):
        "Returns the solution for part one"

        # 1. Return the solution for part one
        numbers = self.get_pair(2020)
        return numbers[0] * numbers[1]

    def part_two(self, verbose=False, limit=0):
        "Returns the solution for part two"

        # 1. Return the solution for part two
        numbers = self.get_trio(2020)
        return numbers[0] * numbers[1] * numbers[2]


# ----------------------------------------------------------------------
#                                                  module initialization
# ----------------------------------------------------------------------
if __name__ == '__main__':
    pass

# ======================================================================
# end                         r e p o r t . p y                      end
# ======================================================================

File: 2020/01_ReportRepair/report.lua
-- ======================================================================
-- Report Repair
--   Advent of Code 2020 Day 01 -- Eric Wastl -- https://adventofcode.com
--
-- lua implementation by Dr. Dean Earl Wright III
-- ======================================================================

-- ======================================================================
--                         r e p o r t . l u a
-- ======================================================================
-- "A solver for the Advent of Code 2020 Day 01 puzzle"

-- ----------------------------------------------------------------------
--                                                                  local
-- ----------------------------------------------------------------------
local Report = {}
-- ----------------------------------------------------------------------
--                                                              constants
-- ----------------------------------------------------------------------

-- ======================================================================
--                                                                  Report
-- ======================================================================

-- "Object for Report Repair"

function Report:Report (o)

  -- 1. Set the initial values
  o = o or {}
  if o.part2 == nil then
    o.part2 = false
  end
  if o.text == nil then
    o.text = {}
  end
  o.numbers = {}

  -- 2. Create the object metatable
  self.__index = self
  setmetatable(o, self)

  -- 3. Process text (if any)
  if o.text ~= nil and #o.text > 0 then
     o:_process_text(o.text)
  end

  -- 4. Return the populated object
  return o
end

function Report:_process_text(text)
  -- "Assign values from text
  
  -- 0. Precondition axioms
  assert(text ~= nil and #text > 0)
  
  -- 1. Loop for each line of the text
  for _, line in ipairs(text) do
    
    -- 2. Add the number to the entries
    table.insert(self.numbers, tonumber(line))
    end  
    
end

function Report:get_pair(total)
  -- "Get a pair of numbers that add to the total"
  for _, num in ipairs(self.numbers) do
    need = total - num
    for _, other in ipairs(self.numbers) do
      if other == need and other ~= num then
        return num, other
      end
    end
  end  
  return nil, nil
end  

function Report:get_trio(total)
  -- "Get a trio of numbers that add to the total"
  for _, num in ipairs(self.numbers) do
    need = total - num
    p1, p2 = self:get_pair(need)
    if p1 ~= nil then
      return num, p1, p2
    end
  end  
  return nil, nil, nil
end  
  
function Report:part_one(args)
  -- "Returns the solution for part one"

  -- 0. Precondition axioms
  local verbose = args.verbose or false
  local limit = args.limit or 0
  assert(verbose == true or verbose == false)
  assert(limit >= 0)

  -- 1. Return the solution for part one
  n1, n2 = self:get_pair(2020)
  if n1 == nil then 
    return nil
  end  
  return n1 * n2
end


function Report:part_two(args)
  -- "Returns the solution for part two"

  -- 0. Precondition axioms
  local verbose = args.verbose or false
  local limit = args.limit or 0
  assert(verbose == true or verbose == false)
  assert(limit >= 0)

  -- 1. Return the solution for part one
  n1, n2, n3 = self:get_trio(2020)
  if n1 == nil then 
    return nil
  end  
  return n1 * n2 * n3
end

-- ----------------------------------------------------------------------
--                                                  module initialization
-- ----------------------------------------------------------------------

return Report

-- ======================================================================
-- end                      r e p o r t . l u a                     end
-- ======================================================================
